packages/material-ui/src/SpeedDial/SpeedDial.test.js

```diff
@@ -154,6 +214,48 @@ describe('<SpeedDial />', () => {
       expect(document.activeElement).to.equal(actions[0]);
       expect(fab).to.have.attribute('aria-expanded', 'true');
     });
+
+    it('should reset the state of the tooltip when the speed dial is closed while it is open', () => {
+      const handleOpen = spy();
+      const { queryByRole, getByRole, getAllByRole } = render(
+        <SpeedDial ariaLabel="mySpeedDial" onOpen={handleOpen}>
+          <SpeedDialAction tooltipTitle="action1" />
+          <SpeedDialAction tooltipTitle="action2" />
+        </SpeedDial>,
+      );
+      const fab = getByRole('button');
+      const menu = getByRole('menu');
+      const actions = getAllByRole('menuitem');
+
+      fab.focus();
+      act(() => {
+        clock.runAll();
+      });
+      expect(menu).to.not.have.class(classes.actionsClosed);
+
+      fireEvent.keyDown(fab, { key: 'ArrowUp' });
+      act(() => {
+        clock.runAll();
+      });
+      expect(queryByRole('tooltip')).not.to.equal(null);
+
+      const escapeEvent = new window.Event('KeyboardEvent');
+      escapeEvent.initEvent('keydown', true);
+      escapeEvent.key = 'Escape';
+      actions[0].dispatchEvent(escapeEvent);
```

Comment:
If `fireEvent.keyDown` is used the test passes even without the [fix](https://github.com/mui-org/material-ui/pull/25259/files#diff-568ef920605c2dd514b8da1ec390e9946f27fb02e3045de295fea37ae9916b39R204). I had to manually dispatch to not have RTL calling it inside an `act`.

Reply:
@m4theushw What test should've failed? I did a CI run without the fix (https://github.com/mui-org/material-ui/pull/25259/commits/34ec9c6053c3c73ae6cfc31e8cf084aa0f0590ab) and [`test_unit`](https://app.circleci.com/pipelines/github/mui-org/material-ui/39095/workflows/429a6d2f-bb84-4429-8d4e-f4851d0f9448/jobs/231037/parallel-runs/0/steps/0-108) and [`test_browser`](https://app.circleci.com/pipelines/github/mui-org/material-ui/39095/workflows/429a6d2f-bb84-4429-8d4e-f4851d0f9448/jobs/231034/parallel-runs/0/steps/0-112) failed as expected

Reply:
> No test should fail

Why shouldn't the test fail? I was under the impression that these illustrate broken behavior the we were trying to fix.

Reply:
> One of my findings was that when `keydown` is dispatched, it updates `open` to false [here](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/SpeedDial/SpeedDial.js?rgh-link-date=2021-03-10T13%3A44%3A34Z#L235), that makes the cleanup function [here](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tooltip/Tooltip.js?rgh-link-date=2021-03-10T13%3A44%3A34Z#L481) to be called. When `act` is used, this [event listener](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tooltip/Tooltip.js?rgh-link-date=2021-03-10T13%3A44%3A34Z#L472) is also called but that doesn't happen in real use or with `dispatchEvent`.

This sounds like another issue of batching vs discrete update semantics. Once we confirmed when tests should fail/what should make them pass, we need to confirm the fix with React 17 as well.

In the end, we might want to use something similar to `fireDiscreteEvent.click`: https://github.com/mui-org/material-ui/blob/HEAD/test/utils/fireDiscreteEvent.js

Reply:
> Why shouldn't the test fail? I was under the impression that these illustrate broken behavior the we were trying to fix.

They should fail if the fix is removed; However, they don't fail if the fix is removed and `fireEvent.keyDown` is used. I demonstrated this in https://github.com/mui-org/material-ui/commit/15987cba7e2c681add53a6b6475a371d5be2b43b.

---

While I was trying to understand why the test was behaving differently from the browser, I found something interesting. If there's a ref callback and this callback changes the state, then the event listeners added in the `useEffect` are not called in the way they should. I created a [codesandbox](https://codesandbox.io/s/material-ui-issue-forked-qo6tx?file=/src/Demo.js) to demonstrate this  behavior. Open it and click the "I am closed" button to attach a "click" event to the document. Click it again to trigger the click listener and the console should log "document click" immediately after the second "button click".  When the ref callback changes the state, the second click is not followed by a "document click" but, instead, it appers at the beginning. To simulate this last situation, uncomment the `setRef` in `handleRef`.

The situation above also happens in the Tooltip. [This](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tooltip/Tooltip.js#L472) event listener is not called correcly because [this](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tooltip/Tooltip.js#L485) ref callback causes a rerender by changing the state. If we managed to make the event listener to be triggered at the right timing this test case would not be necessary.

Ref callback not causing rerender:

![codesandbox-1](https://user-images.githubusercontent.com/42154031/110873152-e5ba1f80-82af-11eb-8804-22c8764f2468.gif)

Ref callback causing rerender:

![codesandbox-2](https://user-images.githubusercontent.com/42154031/110873716-20708780-82b1-11eb-8f17-4245850ffd17.gif)

Reply:
Looks good with `react@next`: https://app.circleci.com/pipelines/github/mui-org/material-ui/39163/workflows/7397fbbf-1f51-447d-9bff-4556a0265d5e

Confirmed that this is yet another case of unwanted batching semantics.

Reply:
> I found something interesting

@m4theushw Strange, indeed. It's similar to https://github.com/facebook/react/issues/20074.

Reply:
> @m4theushw Strange, indeed. It's similar to facebook/react#20074.

Please read my explanation and past explainers regarding discrete event semantics.

Reply:
@eps1lon Oh yeah sorry for the distraction. I was interested in this new thread started in https://github.com/mui-org/material-ui/pull/25259#discussion_r592832198. it's definitely independent of the problem we were initially discussing. I have shared the link to an issue that shows case the same symptoms. Nothing tells it's the same root cause.